samza-core/src/test/java/org/apache/samza/operators/impl/TestOperatorImpl.java

```diff
@@ -18,61 +18,205 @@
+    verify(mockNextOpImpl1, times(1)).handleTimer(mockCollector, mockCoordinator);
+    verify(mockNextOpImpl2, times(1)).handleTimer(mockCollector, mockCoordinator);
+  }
 
   @Test
-  public void testSubscribers() {
-    this.curInputMsg = null;
-    this.curCollector = null;
-    this.curCoordinator = null;
-    OperatorImpl<TestMessageEnvelope, TestOutputMessageEnvelope> opImpl = new OperatorImpl<TestMessageEnvelope, TestOutputMessageEnvelope>() {
-      @Override
-      public void onNext(TestMessageEnvelope message, MessageCollector collector, TaskCoordinator coordinator) {
-        TestOperatorImpl.this.curInputMsg = message;
-        TestOperatorImpl.this.curCollector = collector;
-        TestOperatorImpl.this.curCoordinator = coordinator;
-      }
-      @Override
-      public void onTimer(MessageCollector collector, TaskCoordinator coordinator) {
-
-      }
-
-      };
-    // verify registerNextOperator() added the mockSub and propagateResult() invoked the mockSub.onNext()
-    OperatorImpl mockSub = mock(OperatorImpl.class);
-    opImpl.registerNextOperator(mockSub);
-    TestOutputMessageEnvelope xOutput = mock(TestOutputMessageEnvelope.class);
+  public void testOnTimerUpdatesMetrics() {
+    TaskContext mockTaskContext = mock(TaskContext.class);
+    MetricsRegistry mockMetricsRegistry = mock(MetricsRegistry.class);
+    when(mockTaskContext.getMetricsRegistry()).thenReturn(mockMetricsRegistry);
+    Counter mockMessageCounter = mock(Counter.class);
+    Timer mockTimer = mock(Timer.class);
+    when(mockMetricsRegistry.newCounter(anyString(), anyString())).thenReturn(mockMessageCounter);
+    when(mockMetricsRegistry.newTimer(anyString(), anyString())).thenReturn(mockTimer);
+
+    Object mockTestOpImplOutput = mock(Object.class);
+    OperatorImpl<Object, Object> opImpl = new TestOpImpl(mockTestOpImplOutput);
+    opImpl.init(mock(Config.class), mockTaskContext);
+
+    // send a message to this operator
     MessageCollector mockCollector = mock(MessageCollector.class);
     TaskCoordinator mockCoordinator = mock(TaskCoordinator.class);
-    opImpl.propagateResult(xOutput, mockCollector, mockCoordinator);
-    verify(mockSub, times(1)).onNext(
-        argThat(new IsEqual<>(xOutput)),
-        argThat(new IsEqual<>(mockCollector)),
-        argThat(new IsEqual<>(mockCoordinator))
-    );
-    // verify onNext() is invoked correctly
-    TestMessageEnvelope mockInput = mock(TestMessageEnvelope.class);
-    opImpl.onNext(mockInput, mockCollector, mockCoordinator);
-    assertEquals(mockInput, this.curInputMsg);
-    assertEquals(mockCollector, this.curCollector);
-    assertEquals(mockCoordinator, this.curCoordinator);
+    opImpl.onTimer(mockCollector, mockCoordinator);
+
+    // verify that it updates metrics
+    verify(mockMessageCounter, times(0)).inc();
+    verify(mockTimer, times(1)).update(anyLong());
+  }
+
+  private static class TestOpImpl extends OperatorImpl<Object, Object> {
+    private final Object mockOutput;
```

Comment:
AFAICT, `mockOutput` appears to be unused? Or do some tests assert the behavior that `mockOutput` is actually contained in the emitted result?